homeassistant/components/stream/core.py

```diff
@@ -132,16 +142,20 @@ def get_segments(self) -> deque[Segment]:
         self._idle_timer.awake()
         return self._segments
 
-    async def recv(self) -> Segment | None:
+    async def recv(self, timeout: float | None = None) -> Segment | None:
         """Wait for and retrieve the latest segment."""
-        if self._cursor is None or self._cursor <= self.last_sequence:
-            await self._event.wait()
+        if self._sequence_cursor <= self.last_sequence:
+            try:
+                async with async_timeout.timeout(timeout):
+                    await self._event.wait()
+            except asyncio.TimeoutError:
+                return None
 
         if not self._segments:
             return None
 
-        segment = self.get_segments()[-1]
-        self._cursor = segment.sequence
+        if segment := self.last_segment:
```

Comment:
Yes, you're right..we should probably call self._idle_timer.awake() in recv()

Reply:
> 
> 
> I believe removing the `get_segments` call here a side effective of removing a call to `self._idle_timer.awake()` which keeps the stream from timing out.
> 
> Is that OK because the callers already keep it alive some other way and calling `recv` is optional?
> 
> My impression is that the behavior here is that all the stuff in hls.py keeps the worker alive from the player side perspective from these calls.
> 
> Perhaps doing this as a side effect of reading the segment is a bit too subtle, and instead the "awake" call should be unrolled either into an explicit call or into `recv` here at the top of this function. (I think every caller of `get_segment` or `get_segments` needs to also call `recv` first?)
> 
> I am curious if you have thoughts about if im reading that right, and what might be a good way to address that and/or possibly simplify this in future work.

The repeated render_playlist() calls will keep things alive. But it's two function calls down. We can instead move the `awake` from `get_segments` to `recv` which is more explicit.

Reply:
Oh i think i just remembered now why its this way... I think i tried this and putting it in `recv` may keep recorder alive. This can be tackled later if you get into oddities with tests...

Reply:
Turns out this won't keep it alive since we don't call recv() regularly. I'll add a fix in #51282

Reply:
Would you mind fixing in a separate PR? Can help facilitate quick review

Reply:
Ok, let me undo the commit in #51282